From 976f348b29ff47b64df3c047bd1798986c8ce279 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 16 Apr 2019 14:01:45 +0300
Subject: [PATCH] [SW-0] pll: turn PLLs per device

Each device got its own PLL registers names and nominal settings

Change-Id: Ibe3729ef656c227c9e5890814a08e9bce2a1ed97
---

diff --git a/common/pll.c b/common/pll.c
new file mode 100644
index 0000000..8f65864
--- /dev/null
+++ b/common/pll.c
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2017 HabanaLabs Ltd.
+ *   Author: Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include "preboot.h"
+#include "uart.h"
+#include "pll.h"
+
+static enum hl_pll_bypass_val hl_pll_bypass_get(struct hl_pll_cfg *cfg,
+							enum hl_pll_div_out out)
+{
+	int offset = out * sizeof(u32);
+
+	return regmap_readl(cfg->base + HL_PLL_DIV_SEL + offset);
+}
+
+static void hl_pll_udelay(unsigned int usec)
+{
+	int pll_num;
+	long cnt;
+	struct hl_pll_cfg *hl_pll_cfg;
+	enum hl_pll_bypass_val bval;
+
+	pll_num = hl_pll_get_cfg(&hl_pll_cfg);
+	if (pll_num) {
+		/*
+		 * Assumptions:
+		 * 1) the CPU PLL offset in the array is 0
+		 * 2) DIV1 is the PLL output to timestamp counter.
+		 */
+		bval = hl_pll_bypass_get(hl_pll_cfg, HL_PLL_DIV1);
+	}
+
+	if (bval == HL_PLL_REF_CLK || bval == HL_PLL_DIV_REF)
+		cnt = BYPASS_CYCLES_PER_USEC;
+	else
+		cnt = CPU_CYCELS_PER_USEC;
+
+	cnt = (cnt * usec) / MIN_CYCLES_PER_ITER;
+
+	while (cnt--)
+		;
+}
+
+static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
+{
+	regmap_writel(reset, cfg->base + HL_PLL_RST);
+}
+
+static inline int hl_pll_is_locked(struct hl_pll_cfg *cfg)
+{
+	return regmap_readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
+}
+
+static inline void hl_pll_clear_lock_status(struct hl_pll_cfg *cfg)
+{
+	u32 reg = regmap_readl(cfg->base + HL_PLL_CFG);
+
+	regmap_writel(reg & ~HL_PLL_CFG_LOCK, cfg->base + HL_PLL_CFG);
+}
+
+static int hl_pll_cfg(struct hl_pll_cfg *cfg)
+{
+	int tout = HL_PLL_LOCK_TOUT;
+
+	hl_pll_reset(cfg, true);
+
+	regmap_writel(cfg->nr, cfg->base + HL_PLL_NR);
+	regmap_writel(cfg->nf, cfg->base + HL_PLL_NF);
+	regmap_writel(cfg->od, cfg->base + HL_PLL_OD);
+	regmap_writel(cfg->nb, cfg->base + HL_PLL_NB);
+	regmap_writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
+
+	/* after updating the coefficients we need to wait >=5us */
+	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
+
+	hl_pll_reset(cfg, false);
+
+	/* Wait >500 (10us) reference cycles until PLL re-locks. */
+	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
+
+	/* to be on a safe side, poll on the PLL lock status bit */
+	while (!hl_pll_is_locked(cfg) && tout--)
+		hl_pll_udelay(1);
+
+	if (!tout) {
+		preboot_puts_raw("PLL: Timeout waiting for ");
+		preboot_puts_raw(cfg->name);
+		preboot_puts_raw(" PLL lock...\n\r");
+		return 1;
+	}
+
+	/* w/a the h/w erratum: lock status bit isn't getting cleared */
+	hl_pll_clear_lock_status(cfg);
+
+	return 0;
+}
+
+static void hl_pll_bypass(struct hl_pll_cfg *cfg, bool bypass)
+{
+	int sel;
+	u64 reg;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
+
+		/* -1 means always bypass */
+		if (cfg->div_factor[sel] == -1 || bypass)
+			regmap_writel(HL_PLL_REF_CLK, reg);
+		else
+			regmap_writel(cfg->div_factor[sel] ?
+					HL_PLL_DIV_PLL : HL_PLL_PLL_CLK, reg);
+	}
+
+	/* just to be on a safe side, wait for bypass to propagate */
+	hl_pll_udelay(1);
+}
+
+static void hl_pll_div_init(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u64 div_base;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		if (cfg->div_factor[sel] == -1)
+			continue;
+
+		div_base = cfg->base + (sel * sizeof(u32));
+
+		regmap_writel(cfg->div_factor[sel],
+				div_base + HL_PLL_DIV_FACTOR);
+		regmap_writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
+		regmap_writel(1, div_base + HL_PLL_DIV_EN);
+	}
+
+	/* just to be on a safe side, wait for DIV init to propagate */
+	hl_pll_udelay(1);
+}
+
+static void hl_pll_ack_int(struct hl_pll_cfg *cfg)
+{
+	regmap_writel(1, cfg->base + HL_PLL_LOCK_INTR);
+}
+
+static int hl_pll_set(struct hl_pll_cfg *cfg)
+{
+	int ret;
+
+	hl_pll_bypass(cfg, true);
+
+	ret = hl_pll_cfg(cfg);
+
+	hl_pll_div_init(cfg);
+
+	hl_pll_bypass(cfg, false);
+
+	hl_pll_ack_int(cfg);
+
+	return ret;
+}
+
+/**
+ * hl_pll_init - Initialize PLL registers
+ */
+void hl_pll_init(void)
+{
+	int i, pll_num;
+	struct hl_pll_cfg *hl_pll_cfg;
+	enum hl_pll_bypass_val bval;
+
+	pll_num = hl_pll_get_cfg(&hl_pll_cfg);
+
+	for (i = 0; i < pll_num; i++) {
+		/* DIV0 is the undivided PLL output */
+		bval = hl_pll_bypass_get(hl_pll_cfg + i, HL_PLL_DIV0);
+		if (hl_pll_cfg[i].base == CPU_PLL_BASE &&
+				bval != HL_PLL_REF_CLK &&
+				bval != HL_PLL_DIV_REF)
+			continue;
+
+		hl_pll_set(hl_pll_cfg + i);
+	}
+}
diff --git a/common/pll.h b/common/pll.h
new file mode 100644
index 0000000..d270c70
--- /dev/null
+++ b/common/pll.h
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2017 HabanaLabs Ltd.
+ *   Author: Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#ifndef _HL_PLL_H_
+#define _HL_PLL_H_
+
+#define HL_PLL_NR		0x100
+#define HL_PLL_NF		0x104
+#define HL_PLL_OD		0x108
+#define HL_PLL_NB		0x10c
+#define HL_PLL_CFG		0x110
+#define HL_PLL_CFG_LOCK		(1 << 4)
+#define HL_PLL_LOCK_INTR	0x128
+#define HL_PLL_DATA_CHNG	0x130
+#define HL_PLL_RST		0x134
+#define HL_PLL_DIV_FACTOR	0x200
+#define HL_PLL_DIV_FACTOR_CMD	0x220
+#define HL_PLL_DIV_SEL		0x280
+#define HL_PLL_DIV_EN		0x2a0
+
+#define HL_PLL_MAX_WAIT_TIME	20	/* 20 us */
+#define HL_PLL_LOCK_TOUT	5000	/* 5 ms */
+
+#define USEC_PER_SEC		1000000
+#define CPU_CYCELS_PER_USEC	(1200000000 / USEC_PER_SEC)
+#define BYPASS_CYCLES_PER_USEC	(50000000 / USEC_PER_SEC)
+#define MIN_CYCLES_PER_ITER	5
+
+#define mmPSOC_MME_PLL_BASE	0xC71000
+
+enum hl_pll_bypass_val {
+	HL_PLL_REF_CLK,
+	HL_PLL_PLL_CLK,
+	HL_PLL_DIV_REF,
+	HL_PLL_DIV_PLL,
+};
+
+enum hl_pll_div_out {
+	HL_PLL_DIV0,
+	HL_PLL_DIV1,
+	HL_PLL_DIV2,
+	HL_PLL_DIV3,
+	HL_PLL_DIV_SIZE,
+};
+
+struct hl_pll_cfg {
+	char	*name;
+	u64	base;
+	u32	nr;
+	u32	nf;
+	u32	od;
+	u32	nb;
+	u32	div_factor[HL_PLL_DIV_SIZE];
+};
+
+void hl_pll_init(void);
+int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg);
+
+#endif /* _HL_PLL_H_ */
diff --git a/common/preboot.c b/common/preboot.c
index 3325291..08be6c5 100644
--- a/common/preboot.c
+++ b/common/preboot.c
@@ -16,6 +16,7 @@
 #include "preboot.h"
 #include "uart.h"
 #include "spi.h"
+#include "pll.h"
 
 #define VERSION_MAX_LEN		128
 
diff --git a/gaudi/gaudi.c b/gaudi/gaudi.c
index a401588..5224dce 100644
--- a/gaudi/gaudi.c
+++ b/gaudi/gaudi.c
@@ -9,6 +9,7 @@
 #include <pcie_aux_regs.h>
 
 #include "preboot.h"
+#include "pll.h"
 
 void preboot_pcie_wait_for_link(void)
 {
@@ -31,3 +32,25 @@
 {
 	return mmPSOC_GLOBAL_CONF_CPU_BOOT_STATUS;
 }
+
+static struct hl_pll_cfg gaudi_pll_cfg[] = {
+	{
+		.name = "CPU", .base = mmPSOC_CPU_PLL_BASE,
+		.nr = 0, .nf = 47, .od = 1, .nb = 23,
+		.div_factor = { 0, 1, 47, -1 }
+	}, {
+		.name = "PCI", .base = mmPSOC_PCI_PLL_BASE,
+		.nr = 4, .nf = 331, .od = 3, .nb = 165,
+		.div_factor = { 0, 0, 4, 9 }
+	},
+};
+
+int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg)
+{
+	int pll_num = ARRAY_SIZE(gaudi_pll_cfg);
+
+	*hl_pll_cfg = gaudi_pll_cfg;
+
+	return pll_num;
+}
+
diff --git a/gaudi/gaudi.h b/gaudi/gaudi.h
index 9863004..4ae076f 100644
--- a/gaudi/gaudi.h
+++ b/gaudi/gaudi.h
@@ -8,8 +8,11 @@
 /* Use DumpSel[11] to skip PCIe initialization */
 #define SKIP_PCIE_INIT BIT(19)
 
+#define mmPSOC_CPU_PLL_BASE		0xC70000
+#define mmPSOC_PCI_PLL_BASE		0xC72000
+#define CPU_PLL_BASE			mmPSOC_CPU_PLL_BASE
+
 void preboot_pcie_wait_for_link(void);
 static inline void preboot_delay(void) {}
 static inline void preboot_putc_virt_uart(const char c) {}
-static inline void hl_pll_init(void) {}
 
diff --git a/goya/goya.c b/goya/goya.c
index d8b889b..c0303cc 100644
--- a/goya/goya.c
+++ b/goya/goya.c
@@ -13,6 +13,7 @@
 #include <pcie_db_rsv_regs.h>
 
 #include "preboot.h"
+#include "pll.h"
 
 #define BTL_COUNTER_MAX		BIT(22)
 #define SYS_REF_CLOCK		50000000
@@ -162,3 +163,28 @@
 {
 	return mmPSOC_GLOBAL_CONF_WARM_REBOOT;
 }
+
+static struct hl_pll_cfg goya_pll_cfg[] = {
+	{
+		.name = "CPU", .base = mmCPU_PLL_BASE,
+		.nr = 2, .nf = 159, .od = 1, .nb = 79,
+		.div_factor = { 0, -1, 1, -1 }
+	}, {
+		.name = "IC", .base = mmIC_PLL_BASE,
+		.nr = 0, .nf = 63, .od = 15, .nb = 31,
+		.div_factor = { 0, 1, 1, 2 }
+	}, {
+		.name = "MME", .base = mmPSOC_MME_PLL_BASE,
+		.nr = 0, .nf = 63, .od = 15, .nb = 31,
+		.div_factor = { 0, 1, 2, 8 }
+	},
+};
+
+int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg)
+{
+	int pll_num = ARRAY_SIZE(goya_pll_cfg);
+
+	*hl_pll_cfg = goya_pll_cfg;
+
+	return pll_num;
+}
diff --git a/goya/goya.h b/goya/goya.h
index a0147b0..6b95944 100644
--- a/goya/goya.h
+++ b/goya/goya.h
@@ -8,7 +8,10 @@
 /* Use DumpSel[8] to skip PCIe initialization */
 #define SKIP_PCIE_INIT BIT(21)
 
+#define mmCPU_PLL_BASE		0x4A2000
+#define mmIC_PLL_BASE		0x4A3000
+#define CPU_PLL_BASE		mmCPU_PLL_BASE
+
 void preboot_pcie_wait_for_link(void);
 void preboot_delay(void);
 void preboot_putc_virt_uart(const char c);
-void hl_pll_init(void);
diff --git a/goya/pll.c b/goya/pll.c
deleted file mode 100644
index cf0af4d..0000000
--- a/goya/pll.c
+++ /dev/null
@@ -1,248 +0,0 @@
-#include "preboot.h"
-#include "uart.h"
-
-#define HL_PLL_NR		0x100
-#define HL_PLL_NF		0x104
-#define HL_PLL_OD		0x108
-#define HL_PLL_NB		0x10c
-#define HL_PLL_CFG		0x110
-#define HL_PLL_CFG_LOCK		(1 << 4)
-#define HL_PLL_LOCK_INTR	0x128
-#define HL_PLL_DATA_CHNG	0x130
-#define HL_PLL_RST		0x134
-#define HL_PLL_DIV_FACTOR	0x200
-#define HL_PLL_DIV_FACTOR_CMD	0x220
-#define HL_PLL_DIV_SEL		0x280
-#define HL_PLL_DIV_EN		0x2a0
-
-#define HL_PLL_MAX_WAIT_TIME	20	/* 20 us */
-#define HL_PLL_LOCK_TOUT	5000	/* 5 ms */
-
-#define USEC_PER_SEC		1000000
-#define CPU_CYCELS_PER_USEC	(1200000000 / USEC_PER_SEC)
-#define BYPASS_CYCLES_PER_USEC	(50000000 / USEC_PER_SEC)
-#define MIN_CYCLES_PER_ITER	5
-
-#define mmCPU_PLL_BASE		0x4A2000
-#define mmIC_PLL_BASE		0x4A3000
-#define mmPSOC_MME_PLL_BASE	0xC71000
-
-enum hl_pll_bypass_val {
-	HL_PLL_REF_CLK,
-	HL_PLL_PLL_CLK,
-	HL_PLL_DIV_REF,
-	HL_PLL_DIV_PLL,
-};
-
-enum hl_pll_div_out {
-	HL_PLL_DIV0,
-	HL_PLL_DIV1,
-	HL_PLL_DIV2,
-	HL_PLL_DIV3,
-	HL_PLL_DIV_SIZE,
-};
-
-struct hl_pll_cfg {
-	char	*name;
-	u64	base;
-	u32	nr;
-	u32	nf;
-	u32	od;
-	u32	nb;
-	u32	div_factor[HL_PLL_DIV_SIZE];
-};
-
-static struct hl_pll_cfg goya_pll_cfg[] = {
-	{
-		.name = "CPU", .base = mmCPU_PLL_BASE,
-		.nr = 2, .nf = 159, .od = 1, .nb = 79,
-		.div_factor = { 0, -1, 1, -1 }
-	}, {
-		.name = "IC", .base = mmIC_PLL_BASE,
-		.nr = 0, .nf = 63, .od = 15, .nb = 31,
-		.div_factor = { 0, 1, 1, 2 }
-	}, {
-		.name = "MME", .base = mmPSOC_MME_PLL_BASE,
-		.nr = 0, .nf = 63, .od = 15, .nb = 31,
-		.div_factor = { 0, 1, 2, 8 }
-	},
-};
-
-static int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg)
-{
-	int pll_num = ARRAY_SIZE(goya_pll_cfg);
-
-	*hl_pll_cfg = goya_pll_cfg;
-
-	return pll_num;
-}
-
-static enum hl_pll_bypass_val hl_pll_bypass_get(struct hl_pll_cfg *cfg,
-							enum hl_pll_div_out out)
-{
-	int offset = out * sizeof(u32);
-
-	return regmap_readl(cfg->base + HL_PLL_DIV_SEL + offset);
-}
-
-static void hl_pll_udelay(unsigned int usec)
-{
-	long cnt;
-	/*
-	 * Assumptions:
-	 * 1) the CPU PLL offset in the array is 0
-	 * 2) DIV1 is the PLL output to timestamp counter.
-	 */
-	enum hl_pll_bypass_val bval =
-			hl_pll_bypass_get(goya_pll_cfg, HL_PLL_DIV1);
-
-	if (bval == HL_PLL_REF_CLK || bval == HL_PLL_DIV_REF)
-		cnt = BYPASS_CYCLES_PER_USEC;
-	else
-		cnt = CPU_CYCELS_PER_USEC;
-
-	cnt = (cnt * usec) / MIN_CYCLES_PER_ITER;
-
-	while (cnt--)
-		;
-}
-
-static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
-{
-	regmap_writel(reset, cfg->base + HL_PLL_RST);
-}
-
-static inline int hl_pll_is_locked(struct hl_pll_cfg *cfg)
-{
-	return regmap_readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
-}
-
-static inline void hl_pll_clear_lock_status(struct hl_pll_cfg *cfg)
-{
-	u32 reg = regmap_readl(cfg->base + HL_PLL_CFG);
-
-	regmap_writel(reg & ~HL_PLL_CFG_LOCK, cfg->base + HL_PLL_CFG);
-}
-
-static int hl_pll_cfg(struct hl_pll_cfg *cfg)
-{
-	int tout = HL_PLL_LOCK_TOUT;
-
-	hl_pll_reset(cfg, true);
-
-	regmap_writel(cfg->nr, cfg->base + HL_PLL_NR);
-	regmap_writel(cfg->nf, cfg->base + HL_PLL_NF);
-	regmap_writel(cfg->od, cfg->base + HL_PLL_OD);
-	regmap_writel(cfg->nb, cfg->base + HL_PLL_NB);
-	regmap_writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
-
-	/* after updating the coefficients we need to wait >=5us */
-	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
-
-	hl_pll_reset(cfg, false);
-
-	/* Wait >500 (10us) reference cycles until PLL re-locks. */
-	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
-
-	/* to be on a safe side, poll on the PLL lock status bit */
-	while (!hl_pll_is_locked(cfg) && tout--)
-		hl_pll_udelay(1);
-
-	if (!tout) {
-		preboot_puts_raw("PLL: Timeout waiting for ");
-		preboot_puts_raw(cfg->name);
-		preboot_puts_raw(" PLL lock...\n\r");
-		return 1;
-	}
-
-	/* w/a the h/w erratum: lock status bit isn't getting cleared */
-	hl_pll_clear_lock_status(cfg);
-
-	return 0;
-}
-
-static void hl_pll_bypass(struct hl_pll_cfg *cfg, bool bypass)
-{
-	int sel;
-	u64 reg;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
-
-		/* -1 means always bypass */
-		if (cfg->div_factor[sel] == -1 || bypass)
-			regmap_writel(HL_PLL_REF_CLK, reg);
-		else
-			regmap_writel(cfg->div_factor[sel] ?
-					HL_PLL_DIV_PLL : HL_PLL_PLL_CLK, reg);
-	}
-
-	/* just to be on a safe side, wait for bypass to propagate */
-	hl_pll_udelay(1);
-}
-
-static void hl_pll_div_init(struct hl_pll_cfg *cfg)
-{
-	int sel;
-	u64 div_base;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		if (cfg->div_factor[sel] == -1)
-			continue;
-
-		div_base = cfg->base + (sel * sizeof(u32));
-
-		regmap_writel(cfg->div_factor[sel],
-				div_base + HL_PLL_DIV_FACTOR);
-		regmap_writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
-		regmap_writel(1, div_base + HL_PLL_DIV_EN);
-	}
-
-	/* just to be on a safe side, wait for DIV init to propagate */
-	hl_pll_udelay(1);
-}
-
-static void hl_pll_ack_int(struct hl_pll_cfg *cfg)
-{
-	regmap_writel(1, cfg->base + HL_PLL_LOCK_INTR);
-}
-
-static int hl_pll_set(struct hl_pll_cfg *cfg)
-{
-	int ret;
-
-	hl_pll_bypass(cfg, true);
-
-	ret = hl_pll_cfg(cfg);
-
-	hl_pll_div_init(cfg);
-
-	hl_pll_bypass(cfg, false);
-
-	hl_pll_ack_int(cfg);
-
-	return ret;
-}
-
-/**
- * hl_pll_init - Initialize PLL registers
- */
-void hl_pll_init(void)
-{
-	int i, pll_num;
-	struct hl_pll_cfg *hl_pll_cfg;
-	enum hl_pll_bypass_val bval;
-
-	pll_num = hl_pll_get_cfg(&hl_pll_cfg);
-
-	for (i = 0; i < pll_num; i++) {
-		/* DIV0 is the undivided PLL output */
-		bval = hl_pll_bypass_get(hl_pll_cfg + i, HL_PLL_DIV0);
-		if (hl_pll_cfg[i].base == mmCPU_PLL_BASE &&
-				bval != HL_PLL_REF_CLK &&
-				bval != HL_PLL_DIV_REF)
-			continue;
-
-		hl_pll_set(hl_pll_cfg + i);
-	}
-}
